ollama: Add tool call support

tidely opened the pull request:
The goal of this PR is to support tool calls using ollama. A lot of the serialization work was done in https://github.com/zed-industries/zed/pull/15803 however the abstraction over language models always disables tools. 

## Changelog:

- Use `serde_json::Value` inside `OllamaFunctionCall` just as it's used in `OllamaFunctionCall`. This fixes deserialization of ollama tool calls.
- Added deserialization tests using json from official ollama api docs.
- Fetch model capabilities during model enumeration from ollama provider
- Added `supports_tools` setting to manually configure if a model supports tools

## TODO:

- [x] Fix tool call serialization/deserialization
- [x] Fetch model capabilities from ollama api
- [x] Add tests for parsing model capabilities 
- [ ] Documentation for `supports_tools` field for ollama language model config
- [ ] Convert between generic language model types
- [x] Pass tools to ollama

Release Notes:

- N/A

tidely committed:
ollama: Add deserialization tests

tidely commented:
The best way to allow tool calls on specific models is likely to have the user explicitly allow tools on specific models in `settings.json` under the ollama provider. However this means ollama wouldn't work with the `agent` crate out of the box, rather a user must enable it by allowing tools calls manually.

tidely committed:
add cfg(test) to tests

tidely committed:
simplify test comparison

lj3954 commented:
The [show model information API](https://github.com/ollama/ollama/blob/main/docs/api.md#show-model-information) should have what you're looking for. You'll just have to call it on each model.

tidely commented:
> The [show model information API](https://github.com/ollama/ollama/blob/main/docs/api.md#show-model-information) should have what you're looking for. You'll just have to call it on each model.

Unfortunately I believe the `show` endpoint does not contain a field for whether the model supports tools. Atleast not one which would be consistent across different model families :(

tidely commented:
I just checked with a bunch of different model families are it seems I was wrong! Thank you @lj3954, I'll work on something to make use of it.

tidely committed:
fetch tool support from ollama dynamically

tidely committed:
fix typo

tidely committed:
add test for show model endpoint deserialization

tidely committed:
function to convert to ollama tool

tidely committed:
simplify overly complicated stream filter_map

tidely committed:
return tool completion events when tools are called

tidely committed:
pass tools to ollama

tidely committed:
create unique tool ids

tidely committed:
remove TODO

tidely committed:
remove another filter_map in favor of map

tidely committed:
cleanup

tidely committed:
remove unused role

tidely committed:
map_to_language_model_completion_events

tidely commented on crates/language_models/src/provider/ollama.rs:
```diff
@@ -75,19 +80,29 @@ impl State {
+                .iter()
                 .filter(|model| !model.name.contains("-embed"))
-                .map(|model| ollama::Model::new(&model.name, None, None))
                 .collect();
```
Good catch. I used the length of the filtered Vec to allocate the exact amount of elements in our result Vec, since growing the Vec dynamically would most likely allocate more than the one time we do now by collecting them ahead of time. However I think we can just use the length of the Vec where the elements aren't filtered, and maybe allocate a few elements too much, but we won't have to loop over them twice.

tidely commented on crates/ollama/src/ollama.rs:
```diff
@@ -227,6 +203,18 @@ pub struct ModelDetails {
+
+impl ModelShow {
+    pub fn supports_tools(&self) -> bool {
+        self.capabilities.contains(&"tools".to_string())
```
contains expects a &T, which is &String in this case https://github.com/rust-lang/rust/issues/42671. To the prevent the extra allocation, I think we can use `self.capabilities.iter().any(|x| x == "tools")`

lj3954 replied:
Whoops. I suppose I wasn't patient enough for clippy in a monorepo, I didn't catch that error when I locally made that change.

tidely committed:
remove extra allocation

tidely committed:
remove an extra allocation

lj3954 commented on crates/language_models/src/provider/ollama.rs:
```diff
@@ -75,19 +81,28 @@ impl State {
+                }
+
+                let name = model.name.as_str();
+                // TODO: Explore fetching model capabilities in parallel
```
Is there any case where the uri can point to a remote server? If so, these calls should certainly be parallelized, perhaps with [futures::future::join_all](https://docs.rs/futures/latest/futures/future/fn.join_all.html). Otherwise, there's no benefit.

tidely replied:
I'll implement it now, but we should consider revisiting this and adding a rate limit. There might be an arbitrary amount of models available, so we would create a big burst of requests.

tidely replied:
Creating a rate limit turned out to be trivial using `futures_util::stream::stream::buffer_unordered::BufferUnordered`, sending a commit in shortly.

tidely committed:
parallelize model capability fetching

tidely committed:
rate-limit capability fetching

tidely committed:
use ToString trait on Value over serde_json::to_string

Value implements ToString, so we dont need to pass it through serde_json::to_string. This saves us from handling an error that can never occur since serde_json::json! always returns a valid Value

tidely commented:
@mgsloan Hey! I saw you recently worked on https://github.com/zed-industries/zed/pull/29885. I looked at creating a similar map_to_completion_events function for the ollama provider, but it doesn't seem to be working as I hoped for. Would you like to take a look at this PR's implementation and see if there's anything you could point out at a quick glance? Otherwise I'll schedule myself a deep dive into the agent crate. Thank you in advance!

as-cii committed:
Merge remote-tracking branch 'origin/main' into ollama-tools

as-cii committed:
ollama: Stop with a `ToolUse` reason when there are pending tool uses

Co-authored-by: Nathan Sobo <nathan@zed.dev>

nathansobo commented:
Thanks so much for getting this started! Very helpful :heart:

as-cii commented:
Really nice work here, @tidely! We made some small tweaks to your pull request, but it was already working great for the most part!
